# TM10007 Assignment template

In [3]:
# Run this to use from colab environment
%pip install -q --upgrade git+https://github.com/jveenland/tm10007_ml.git

Note: you may need to restart the kernel to use updated packages.


## Data loading and cleaning

Below are functions to load the dataset of your choice. After that, it is all up to you to create and evaluate a classification method. Beware, there may be missing values in these datasets. Good luck!

In [4]:
# Data loading functions. Uncomment the one you want to use
from worcliver.load_data import load_data
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

data = load_data()

# Data clean-up (before splitting data)
In this section double entries, such as repeated features and samples, will be removed from the dataset.

In [5]:
df = pd.DataFrame(data)

'''
These 2 lines add a duplicate sample in the DataFrame, you can than observe what the effect is 
in the printed sentence. If you'd like to see it indeed works, please uncomment these lines.
'''
# row = df.iloc[0].copy()
# df = df.append(row, ignore_index=True)
'''
In the lines below we first drop all duplicate entries, whether it is a duplicate row or a column. 
Additionally, the number of duplicates will be returned to the user.
'''

df_f = df.drop_duplicates()
print(f'The number of samples (after removal of duplicates): {len(df_f.index)}, the number of duplicates that had to be removed: {len(df.index)-len(df_f.index)}')
print(f'The number of columns (after removal of duplicates): {len(df_f.columns)}, the number of duplicates that had to be removed: {len(df.columns)-len(df_f.columns)}')

The number of samples (after removal of duplicates): 186, the number of duplicates that had to be removed: 0
The number of columns (after removal of duplicates): 494, the number of duplicates that had to be removed: 0


In [6]:
df = df_f                                       # So we can just use df from now on
df = df.replace('', np.nan)                     # Any empty values are replaced for NaN
# df.describe(include='all')                    # Uncomment if interested

### Conclusion
The table above shows there are only 2 labels present in the dataset, from observation of the csv.file we can see the labels are 'malignant' and 'benigne'. What is also stated here, that 94 of the 186 samples was labeled malignant (and thus 92 samples are labeled benigne.) This information is usefull in the next step, to determine whether a stratified splitting of our data, would be necessary.

# Data splitting (1/2)
In this section the dataset will be split into two datasets, the Design set (D_set) and the Final Test set (Ft_set). This splitting will be performed randomly where 80% will end up in D_set and the remaining 20% will be stored in Ft_set. Ft_set will not be used at all, until the very latest to evaluate our tool and establish how generalising the tool is. In the initial split we will randomly split the dataset, as both labels are equally abundant in the set.

In [7]:
from sklearn.model_selection import train_test_split
# D_set, Ft_set = train_test_split(df, test_size=0.2)
# D_set = pd.DataFrame(D_set)
# Ft_set = pd.DataFrame(Ft_set)

# D_set.describe(include='all')                             

In [8]:
# D_set.to_csv("C:/Users/dgjpa/D_set.csv")
# Ft_set = Ft_set.to_csv("C:/Users/dgjpa/Ft_set.csv")

'''
These lines were used to save the splitted sets, to make sure Design and Test set won't alter.
''' 

"\nThese lines were used to save the splitted sets, to make sure Design and Test set won't alter.\n"

# Data splitting (2/2)
In this section D_set will be split into two datasets, the train set (Tr_set) and the Validation set (Va_set). This splitting will be performed randomly where 90% will end up in Tr_set and the remaining 10% will be stored in Va_set. This second splitting will be repeated 10 times with each iteration a different 10% of the samples will be put into the Va_set, as we will use cross-validation to make most out of the data available. 

In [9]:
# Load D_set here and start second split.